### ISSUE TITLE
TigerVNC 1.6.0 client FullScreenAllMonitors doesn't

### ISSUE THREAD

**Issue by tken3** [TigerVNC/tigervnc#228]
Using the binary release for client and server at:
  https://bintray.com/artifact/download/tigervnc/stable/tigervnc-Linux-x86_64-1.5.0.tar.gz

Selecting options: 
  FullScreenAllMonitors=1 FullScreen=1

On machines with multiple monitors, fullscreen only expands to 1 monitor. FullScreenAllMonitors option has no effect.

Observed this behavior with 1.5.0 client (from bintray link above) running on CentOS 6 and Mint 17.2, connecting to 1.5.0 or 1.4.3 server.

Just switching the client to TigerVNC 1.4.3, FullScreenAllMonitors works as expected.

To summarize:

| Client | Server | FullScreenAllMonitors |
| --- | ---- | ---- |
| 1.4.3 | 1.5.0 | Works |
| 1.5.0, 1.6.0 | 1.4.3 | 1 monitor only |
| 1.5.0, 1.6.0  | 1.5.0 | 1 monitor only |

**Comment by CendioOssman** [comment]
@bphinz ? I'm not getting this problem with my build, only the official one.

**Comment by bphinz** [comment]
How about the latest nightly (which is currently out of date, working on
getting those online again...)?

-brian

**Comment by CendioOssman** [comment]
Same problem with the nightlies unfortunately.

**Comment by bphinz** [comment]
I looked through the build logs and didn't see anything that jumped out at
me. There was one warning that I believe only showed up on the el5 and el6
builds, but I don't know if it's related or even meaningful. I'll send you
the logs and maybe you can take a look.

-brian

**Comment by CendioOssman** [comment]
Is there any way to get a build with debug symbols? That way I could use gdb and hopefully dig a bit more.

**Comment by tken3** [comment]
Same behavior with 1.6.0 client.

**Comment by CendioOssman** [comment]
I assume you didn't mean to close it. :)

I can confirm that 1.6.0 still suffers from this. Can't test the nightly as it crashes for me.

**Comment by CendioOssman** [comment]
New nightlies tested. No more crash, but still no multi-head. The problem is that your FLTK is being built without Xinerama support. So @bphinz, you need to check your build logs and see why FLTK isn't picking that up properly.

**Comment by bphinz** [comment]
On Tue, Mar 1, 2016 at 8:16 AM, Pierre Ossman (Work account) <
notifications@github.com> wrote:

> New nightlies tested. No more crash, but still no multi-head. The problem
> is that your FLTK is being built without Xinerama support. So @bphinz
> <https://github.com/bphinz>, you need to check your build logs and see
> why FLTK isn't picking that up properly.
>
It's missing a link against Xext.  I've fixed that but found a couple of
others.  I should have new binaries ready sometime tomorrow.

-brian

**Comment by bphinz** [comment]
On Tue, Mar 1, 2016 at 11:16 PM, Brian Hinz wrote:

> On Tue, Mar 1, 2016 at 8:16 AM, Pierre Ossman (Work account) <
> notifications@github.com> wrote:
>
>> New nightlies tested. No more crash, but still no multi-head. The problem
>> is that your FLTK is being built without Xinerama support. So @bphinz
>> <https://github.com/bphinz>, you need to check your build logs and see
>> why FLTK isn't picking that up properly.
>>
> It's missing a link against Xext.  I've fixed that but found a couple of
> others.  I should have new binaries ready sometime tomorrow.
>

Can you guys test the newest nightlies and see if the issue is resolved?

**Comment by CendioOssman** [comment]
Still broken. Please check that HAVE_XINERAMA has been set in the FLTK build, otherwise something is still wrong.

**Comment by bphinz** [comment]
Grrr, static linking is such a pain in the ass... HAVE_XINERAMA is set in
the FLTK build, and I don't see any other issues in the config log there.
I'll look over the build logs for vncviewer tonight.

**Comment by CendioOssman** [comment]
I don't think the linking is the problem. I'm looking at the disassembled code (debug symbols in FLTK would be nice \*wink\* \*nudge\*), and I'm not seeing any trace of the code inside `#ifdef HAVE_XINERAMA`. So I'd say ignore the linking for now. There is something more fundamental going wrong.

**Comment by CendioOssman** [comment]
So I'm trying to build this myself, but your EL5 build is just too magical. So far I've had to grab autoconf and automake from EL6 (which there is a comment for, but still). nasm is missing as a dependency and the one in EL5 seems to be too old. For now I disabled SIMD in libjpeg. But then libdrm goes poof with some weird libtool error.

This RPM may run fine on EL5, but it sure doesn't build fine there. :/

**Comment by bphinz** [comment]
On Fri, Mar 4, 2016 at 7:41 AM, Pierre Ossman (Work account) <
notifications@github.com> wrote:

> So I'm trying to build this myself, but your EL5 build is just too
> magical. So far I've had to grab autoconf and automake from EL6 (which
> there is a comment for, but still). nasm is missing as a dependency and the
> one in EL5 seems to be too old. For now I disabled SIMD in libjpeg. But
> then libdrm goes poof with some weird libtool error.
>
Yeah, unfortunately el5 just can't satisfy some of the dependencies on it's
own.  I also used autoconf & automake from el6, and I guess maybe the same
goes for nasm (I'll update that dependency).  Not sure about the
libdrm/libtool error.  I tried migrating the build to el6, which would
solve a lot of these issues, but IIRC the resulting binaries would not run
on el5 because it wanted a newer GLIBC (? at I think that's what the
problem was, it's been quite a while...)

> This RPM may run fine on EL5, but it sure doesn't build fine there. :/
>
I'm certainly open to suggestions on how we can fix this abomination.  IMO,
the only reason for keeping support for el5 at all is to produce the
generic binaries.

I dug into how to make rpmbuild create debug info packages for a
sub-package (in this case the static-devel package) but it appears
surprisingly obscure.

**Comment by CendioOssman** [comment]
> I tried migrating the build to el6, which would solve a lot of these issues, but IIRC the resulting binaries would not run on el5 because it wanted a newer GLIBC (? at I think that's what the problem was, it's been quite a while...)

That is expected. You will generally be bound to the glibc version you build against (or newer). But RHEL 5 is ancient, which is evident by how many things you need to back port. Perhaps raising the minimum requirements to EL6 is not that wrong?

> I also used autoconf & automake from el6, and I guess maybe the same goes for nasm (I'll update that dependency)

I tried grabbing nasm from EL6, but it requires EL6's newer glibc. So no dice. Needs at the very least a recompile.

> I'm certainly open to suggestions on how we can fix this abomination. IMO, the only reason for keeping support for el5 at all is to produce the generic binaries.

The generic binaries are good. But the EL5 target is being excessively compatible (IMO). So keep the principle of linking most stuff statically, but do so on EL6 where we can satisfy the tool requirements.

> I dug into how to make rpmbuild create debug info packages for a sub-package (in this case the static-devel package) but it appears surprisingly obscure.

Static linking is frowned upon, so it's generally ignored. The tools will ignore .a files so you need to have .so files as well to get the source packaged up. **However**, the debug symbols are therefore also never stripped from the .a files. Meaning they end up in whatever you link them in to. So just making sure `-g` is used for the FLTK build will go a long way.

**Comment by bphinz** [comment]
I've added the NASM build prereq to the spec file and added '-g' to CFLAGS if you want to look at the debug symbols again.  I still don't see why HAVE_XINERAMA is showing up in FLTK, because configure is definitely finding it:

...
checking for glXMakeCurrent in -lGL... no
checking for glXMakeCurrent in -lMesaGL... no
checking for glXGetProcAddressARB in -lGL... no
checking GL/glu.h usability... yes
checking GL/glu.h presence... yes
checking for GL/glu.h... yes
checking for XineramaIsActive in -lXinerama… yes
…

-brian

**Comment by CendioOssman** [comment]
> I've added the NASM build prereq to the spec file and added '-g' to CFLAGS if you want to look at the debug symbols again. I still don't see why HAVE_XINERAMA is showing up in FLTK, because configure is definitely finding it:

You put the BuildRequires in the wrong place in the spec file so it's not having any real effect. Please try to test your stuff before you push it to master.

I'm not sure where to get that version of nasm on to EL5. Did you rebuild the SRPM from EL6?

**Comment by CendioOssman** [comment]
> I'm not sure where to get that version of nasm on to EL5. Did you rebuild the SRPM from EL6?

And I just noticed your comment on the commit...

**Comment by CendioOssman** [comment]
We have new nightly builds up that work fine for me. @tken3, could you test out the generic Linux builds?

**Comment by tken3** [comment]
On Tue, 2016-03-29 at 03:57 -0700, Pierre Ossman (Work account) wrote:
> We have new nightly builds up that work fine for me. @tken3, could you
> test out the generic Linux builds?

The nightly build works great!
http://tigervnc.bphinz.com/nightly/xc/x86_64/tigervnc-1.6.80-20160329git113d82f.x86_64.tar.gz

**Comment by CendioOssman** [comment]
Great. Considering this issue closed then.
